oauthproxy.go

```diff
@@ -1145,12 +1153,12 @@ func validOptionalPort(port string) bool {
 // - `nil, ErrAccessDenied` if the authenticated user is not authorized
 // Set-Cookie headers may be set on the response as a side-effect of calling this method.
 func (p *OAuthProxy) getAuthenticatedSession(rw http.ResponseWriter, req *http.Request) (*sessionsapi.SessionState, error) {
-	var session *sessionsapi.SessionState
+	session := middlewareapi.GetRequestScope(req).Session
 
-	getSession := p.sessionChain.Then(http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
-		session = middlewareapi.GetRequestScope(req).Session
-	}))
-	getSession.ServeHTTP(rw, req)
+	// Check this after loading the session so that if a valid session exists, we can add headers from it
+	if p.IsAllowedRequest(req) {
+		return session, nil
+	}
```

Comment:
Do we need to validate that the session is also valid before we return it? Could this be a security hole if we return what would otherwise be an invalid session? Eg, if someones IP allows them through but their session does not have the correct group requirements? Should the session be passed here or not? (CC @ianroberts @NickMeves)

Reply:
I can't remember where this got fixed? I vaguely recall an issue asking for adding session headers to skipped routes if a session exists and doing nothing otherwise? Did we implement that?

Either way -- I think even if the AuthZ logic on session fields would fail, we still pass it to the backend if they are under a `IsAllowedRequest`. People have wanted the `X-Forwarded-*` auth headers to perform their own optional AuthZ logic on the backend when they need to share the allowed route with some external/API requests that can't auth.

Reply:
Or is this us implementing that functionality for the first time? 😄

Reply:
This is us doing it for the first time, but yeah, there have been some issues discussing it!

I do wonder  whether it is a security issue to be passing an invalid session to the backend if they have an allowed request.

@ianroberts was one of the people involved in the original issue about injecting this, perhaps he has a moment to chime in? 

If the session is invalid (ie doesn't match the email validation or has an expired token or something like that), do we expect the headers to be injected or do we expect the session to be ignored as it doesn't pass authz?

Reply:
Actually, the more I think about it, it should be a valid session as we don't put it into the scope unless it's valid, so I think this is safe as is. I think this PR is actually ok to merge, but we should wait until the release first

Reply:
> we don't put it into the scope unless it's valid

Indeed, that's my reading of the various session loader middlewares too - `scope.Session = ...` is always the last thing the handler functions do before delegating to the next handler in the chain.

Reply:
P.S. @JoelSpeed I've implemented a bit of logic at https://github.com/ianroberts/oauth2-proxy/tree/proxy-router on top of this PR to suppress the various "couldn't load session" log messages in the case where the session is optional (i.e. the route is skipped), if that's of interest.

Reply:
> Actually, the more I think about it, it should be a valid session as we don't put it into the scope unless it's valid, so I think this is safe as is. I think this PR is actually ok to merge, but we should wait until the release first

Yeah - the `Authorize` method for each provider is called as the final step to the sign-in flow before a session is created (in addition to subsequently on every request). That was missing on the initial implementation, but caused security/logic issues for the subrequest flow if I recall?

But yes - our sessions should all be valid.